presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java

```diff
@@ -243,6 +249,16 @@ private boolean queryCreated(DispatchQuery dispatchQuery)
             stats.trackQueryStats(dispatchQuery);
         }
 
+        // capture failures which happen before query is succesfully dispatched.
+        dispatchQuery.addStateChangeListener(newState -> {
+            if (newState == FAILED) {
+                if (!dispatchQuery.getDispatchedFuture().isDone() || getDone(dispatchQuery.getDispatchedFuture()) == DispatchQuery.DispatchStatus.FAILED) {
```

Comment:
i'm concerned about `!dispatchQuery.getDispatchedFuture().isDone()` part. This may result in double delivery of the event. 
We could perhaps have the `queryMonitor.queryImmediateFailureEvent` run as a completion callback on the `dispatchQuery.getDispatchedFuture()` future?

Reply:
Let me see. I think it can be done.

Reply:
With code like this:
```java
        // capture failures which happen before query is successfully dispatched.
        addSuccessCallback(dispatchQuery.getDispatchedFuture(),
                dispatchStatus -> {
                    if (dispatchStatus == DispatchQuery.DispatchStatus.FAILED) {
                        queryMonitor.queryImmediateFailureEvent(dispatchQuery.getBasicQueryInfo(), dispatchQuery.getFullQueryInfo().getFailureInfo());
                    }
                });
```

it seems to work fine. 
I will read code some more to see if I see some issues with it.

Reply:
@findepi I guess we should not use the same-thread-executor here, right? So with this change I would need to provide extra executor do `DispatchManager`. Does it still make sense to you?

Reply:
I can also leave original callback on state change, and change the `if` to:
```
if (dispatchQuery.getDispatchedFuture().isDone() && getDone(dispatchQuery.getDispatchedFuture()) == DispatchQuery.DispatchStatus.FAILED) {
```
This captures  the intent more, but depends on the fact that listeners are called in registration order. That listner which finalizes `disaptchedFuture` would be called before this one.
Given fact that listeners are called asynchronously in separate threads I think this is not the thing we can assume.

It feels to me that you proposal with listener on `getDispatchedFuture` completion is most bullet proof.